### PR TITLE
feat(Update Swimm-UI to typescript): Updated typescript support

### PR DESCRIPTION
**Pull request by yhamry**
Added typescript
Changed all components to script setup and lang=ts Moved all constants to files
Added tsconfig

### CODE REVIEW

**Comment by talkor** [comment]
@segevfiner can you review the TS configs (and pnpm) please?

File: .eslintrc.cjs [hunk 1]
```diff
@@ -7,6 +7,7 @@ module.exports = {
     'plugin:vue/vue3-essential',
     'eslint:recommended',
     '@vue/eslint-config-prettier',
+    '@vue/eslint-config-typescript/recommended',
```
Comment (segevfiner) [inline .eslintrc.cjs:10]: Those two are sensitive to order. make sure the order is right as it is in `create-vue`

File: src/components/Action/Action.vue [hunk 2]
```diff
@@ -1,3 +1,89 @@
+<script setup lang="ts">
```
Comment (segevfiner) [inline src/components/Action/Action.vue:1]: You might need to disable this rule lint rule

File: src/components/Action/Action.vue [hunk 3]
```diff
@@ -1,3 +1,89 @@
...
+   * @values big, small
+   */
+  size: {
+    type: String,
+    default: SIZES.BIG,
+    validator: (value: string) => Object.values(SIZES).includes(value),
+  },
+  /**
+   * Flag for secondary button style
+   * @type {boolean}
```
Comment (segevfiner) [inline src/components/Action/Action.vue:19]: You can get rid of `@types`.

File: src/components/Action/Action.vue [hunk 4]
```diff
@@ -1,3 +1,89 @@
+<script setup lang="ts">
+import { computed } from 'vue';
+import Icon from '../Icon/Icon.vue';
+import { VARIANTS, SIZES, TYPES } from './constants';
+
+defineEmits(['click']);
+const props = defineProps({
+  /**
+   * The size of the button. Defaults to big.
+   * @values big, small
```
Comment (segevfiner) [inline src/components/Action/Action.vue:10]: The correct way to do this with TS is define `type: String as PropType<Sizes>`, where `Sizes` is an enum. in typescript, it ignores JSDoc/TSDoc comments like this AFAIK. Probably in many other places.

File: src/components/Action/constants.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export const SIZES = { BIG: 'big', SMALL: 'small' };
```
Comment (segevfiner) [inline src/components/Action/constants.ts:1]: The correct way is to use typescript enums string union depeneding on how you want it used.

File: tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    "allowJs": true,
+    "strict": false,
+    "alwaysStrict": true,
+    "noImplicitReturns": true,
+    "noImplicitThis": true,
+    "noImplicitAny": true,
+    "strictBindCallApply": true,
+    "strictFunctionTypes": true,
+    "allowUnreachableCode": false,
+    "noImplicitOverride": true,
```
Comment (segevfiner) [inline tsconfig.json:15]: Not fully `strict` 😢

File: tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    "paths": {
+      "@/*": ["./src/*"]
+    }
+  },
+  "include": ["src/**/*", "src/**/*.vue"],
```
Comment (segevfiner) [inline tsconfig.json:20]: You might be missing `vite`'s `env.d.ts`, check a `create-vue` template.

File: vite.config.js [hunk 8]
```diff
@@ -15,9 +23,9 @@ export default defineConfig({
   },
   build: {
     lib: {
-      entry: path.resolve(__dirname, 'src/index.js'),
+      entry: path.resolve(__dirname, 'src/index.ts'),
       name: 'Swimm UI',
-      fileName: (format) => `swimm-ui.${format}.js`,
+      fileName: (format) => `swimm-ui.${format}.ts`,
```
Comment (segevfiner) [inline vite.config.js:28]: That's not right. The output is supposed to be a JS file. Consider if we really a UMD build, and set `formats` accodrdingly.

File: vite.config.js [hunk 9]
```diff
@@ -15,9 +23,9 @@ export default defineConfig({
...
     },
     rollupOptions: {
       external: ['vue'],
```
Comment (segevfiner) [inline vite.config.js:31]: Make sure this is complete and there are no external deps bundled into the bundle that shouldn't be.

File: yarn.lock [hunk 10]
```diff
@@ -281,6 +281,48 @@
     "@jridgewell/resolve-uri" "^3.0.3"
```
Comment (segevfiner) [inline yarn.lock:281]: Delete `yarn.lock` if you go `pnpm`.

File: package.json [hunk 11]
```diff
@@ -33,7 +33,9 @@
     "@semantic-release/git": "^10.0.1",
```
Comment (segevfiner) [inline package.json:33]: Need to add `packageManager` and `engines` for proper pnpm with corepack.

File: src/components/Breadcrumbs/Breadcrumb/types.ts [hunk 12]
```diff
@@ -0,0 +1,5 @@
+export interface BreadcrumbType {
```
Comment (segevfiner) [inline src/components/Breadcrumbs/Breadcrumb/types.ts:1]: You can actually export types from `vue` files even while using `script setup` by exporting them from a normal `script lang="ts"` block. Consider this for other places as well

**Comment by segevfiner** [comment]
Another thing is that you also need to update the `package.json` as appropriate so that the types are properly exposed, check one of our other projects that uses `vite-plugin-dts` for an example.

File: .eslintrc.cjs (hunk 1 above)
Comment (yhamry) [reply .eslintrc.cjs:10]: I'll switch them as it is in Swimm app

File: package.json (hunk 11 above)
Comment (yhamry) [reply package.json:33]: Currently haven't updated to pnpm on swimm-ui @talkor WDYT

File: src/components/Action/Action.vue (hunk 2 above)
Comment (yhamry) [reply src/components/Action/Action.vue:1]: Yeah I guess

File: src/components/Action/Action.vue (hunk 4 above)
Comment (yhamry) [reply src/components/Action/Action.vue:10]: cool - Done

File: src/components/Action/Action.vue (hunk 3 above)
Comment (yhamry) [reply src/components/Action/Action.vue:19]: Fixed all JSDocs

File: tsconfig.json (hunk 6 above)
Comment (yhamry) [reply tsconfig.json:15]: Changed `strict` to true

File: tsconfig.json (hunk 7 above)
Comment (yhamry) [reply tsconfig.json:20]: Didn't see any
https://github.com/vitejs/vite/blob/main/packages/create-vite/template-vue-ts/tsconfig.json

Maybe on `tsconfig.node.json`? as in [here](https://github.com/vitejs/vite/blob/main/packages/create-vite/template-vue-ts/tsconfig.node.json)

File: vite.config.js (hunk 9 above)
Comment (yhamry) [reply vite.config.js:31]: As far as I know
@talkor ?

**Review by yhamry — COMMENTED** [review]
Thanks! @segevfiner

File: vite.config.js (hunk 9 above)
Comment (talkor) [reply vite.config.js]: Yeah I think @segevfiner has a point, as long as it's a dependency I think it's fine

**Comment by yhamry** [comment]
@segevfiner - Done!
Please review and approve when you can

File: .github/workflows/lint.yml [hunk 13]
```diff
@@ -10,20 +10,23 @@ jobs:
       - uses: actions/setup-node@v2
         with:
           node-version: '16'
+      - uses: pnpm/action-setup@v2.2.4
```
Comment (segevfiner) [inline .github/workflows/lint.yml:13]: Should be before node

File: .github/workflows/release.yml [hunk 14]
```diff
@@ -15,32 +15,37 @@ jobs:
       - uses: actions/setup-node@v2
         with:
           node-version: '16'
+
+      - uses: pnpm/action-setup@v2.2.4
```
Comment (segevfiner) [inline .github/workflows/release.yml:19]: Should be before node

File: .github/workflows/test.yml [hunk 15]
```diff
@@ -10,20 +10,23 @@ jobs:
       - uses: actions/setup-node@v2
         with:
           node-version: '16'
+      - uses: pnpm/action-setup@v2.2.4
```
Comment (segevfiner) [inline .github/workflows/test.yml:13]: Should be before setup-node

File: src/components/Avatar/SwAvatar.vue [hunk 16]
```diff
@@ -1,48 +1,32 @@
+<script setup lang="ts">
+import { computed } from 'vue';
+import { SHAPE, SIZE } from './constants';
+
+const props = defineProps({
+  text: { type: String, required: true },
+  src: { type: String, default: '' },
+  size: {
+    type: String,
```
Comment (segevfiner) [inline src/components/Avatar/SwAvatar.vue:9]: PropType?

File: src/components/Action/constants.ts [hunk 17]
```diff
@@ -0,0 +1,12 @@
...
+  SMALL = 'small',
+}
+export enum VARIANTS {
+  DANGER = 'danger',
+  SUCCESS = 'success',
+}
+export enum TYPES {
+  BUTTON = 'button',
+  LINK = 'a',
+}
```
Comment (segevfiner) [inline src/components/Action/constants.ts:12]:
```suggestion
export enum SIZES {
  BIG = 'big',
  SMALL = 'small',
}

export enum VARIANTS {
  DANGER = 'danger',
  SUCCESS = 'success',
}

export enum TYPES {
  BUTTON = 'button',
  LINK = 'a',
}
```

File: src/components/Avatar/constants.ts [hunk 18]
```diff
@@ -0,0 +1,13 @@
+export const SHAPE = {
```
Comment (segevfiner) [inline src/components/Avatar/constants.ts:1]: Should probably be enums

File: src/components/Breadcrumbs/Breadcrumbs.vue [hunk 19]
```diff
@@ -1,3 +1,20 @@
+<script setup lang="ts">
+import Breadcrumb from './Breadcrumb/Breadcrumb.vue';
+import type { BreadcrumbType } from './Breadcrumb/types';
+
+defineProps({
+  items: {
+    type: Array as () => Array<BreadcrumbType>,
```
Comment (segevfiner) [inline src/components/Breadcrumbs/Breadcrumbs.vue:7]: I think you are supposed to use `PropType`

File: src/components/Toggle/SwToggle.vue [hunk 20]
```diff
@@ -1,8 +1,30 @@
+<script setup lang="ts">
+import { ref } from 'vue';
+import { SIZE } from './constants';
+
+const props = defineProps({
+  value: { type: Boolean, required: true },
+  disabled: { type: Boolean, default: false },
+  size: {
+    type: String,
```
Comment (segevfiner) [inline src/components/Toggle/SwToggle.vue:9]: `PropType`

File: src/components/Toggle/constants.ts [hunk 21]
```diff
@@ -0,0 +1,6 @@
+export const SIZE = {
```
Comment (segevfiner) [inline src/components/Toggle/constants.ts:1]: Should be an enum

File: tsconfig.json (hunk 7 above)
Comment (segevfiner) [reply tsconfig.json:20]: Some files are generated by code. Do `pnpm create vue@3` and check it out.

File: .github/workflows/lint.yml (hunk 13 above)
Comment (yhamry) [reply .github/workflows/lint.yml:10]: Done

File: .github/workflows/release.yml (hunk 14 above)
Comment (yhamry) [reply .github/workflows/release.yml]: Done

File: src/components/Avatar/constants.ts (hunk 18 above)
Comment (yhamry) [reply src/components/Avatar/constants.ts]: Done

File: src/components/Breadcrumbs/Breadcrumbs.vue (hunk 19 above)
Comment (yhamry) [reply src/components/Breadcrumbs/Breadcrumbs.vue]: Done

File: src/components/Toggle/SwToggle.vue (hunk 20 above)
Comment (yhamry) [reply src/components/Toggle/SwToggle.vue]: Done

File: src/components/Toggle/constants.ts (hunk 21 above)
Comment (yhamry) [reply src/components/Toggle/constants.ts]: Done

**Comment by yhamry** [comment]
@segevfiner - Look at the files from vite-create

File: .github/workflows/release.yml [hunk 22]
```diff
@@ -12,35 +12,40 @@ jobs:
...
           fetch-depth: 0
           token: ${{ secrets.GH_CI_TOKEN }}
+      - uses: pnpm/action-setup@v2.2.4
+        with:
+          version: 7
       - uses: actions/setup-node@v2
         with:
           node-version: '16'
+
+
```
Comment (yhamry) [inline .github/workflows/release.yml:22]: REmove
```suggestion
```

**Review by talkor — APPROVED** [review]

**Pull request merged**